lib/tests/sources.sh

```diff
@@ -0,0 +1,322 @@
+
+
+dir="$(
+  nix eval --raw '(with import <nixpkgs/lib>; with sources; "${
+    extend
+      (cleanSourceWith { src = cleanSource ./.;
+                         filter = path: type: baseNameOf path != "src"; })
+      [ ./src ]
+  }")'
+)"
+(cd "$dir"; find .) | sort -f | diff -U10 - <(cat <<EOF
+.
+./foo.bar
+./README.md
+./src
+./src/bar.c
+./src/bar.o
+EOF
+) || die "sources.extend: filtered lhs"
+
+dir="$(
+  nix eval --raw '(with import <nixpkgs/lib>; with sources; "${
+    extend
+      (cleanSourceWith { src = cleanSource ./.;
+                         filter = path: type: baseNameOf path != "src"; })
+      [ (cleanSourceWith { src = ./src;
+                         filter = path: type: !hasSuffix ".c" path; }) ]
+  }")'
+)"
+(cd "$dir"; find .) | sort -f | diff -U10 - <(cat <<EOF
+.
+./foo.bar
+./README.md
+./src
+./src/bar.o
+EOF
+) || die "sources.extend: lhs filter doesn't influence rhs files"
+
+
+dir="$(
+  nix eval --raw '(with import <nixpkgs/lib>; with sources; "${
+    extend
+      (cleanSourceWith { src = reparent ./. ./src;
+                         filter = path: type: !hasSuffix ".c" path; })
+      [ (cleanSourceWith { src = cleanSource ./.;
+                         filter = path: type: baseNameOf path != "src"; }) ]
+  }")'
+)"
+(cd "$dir"; find .) | sort -f | diff -U10 - <(cat <<EOF
+.
+./foo.bar
+./README.md
+./src
+./src/bar.o
+EOF
+) || die "sources.extend: rhs filter doesn't influence lhs files"
+
+dir="$(
+  nix eval --raw '(with import <nixpkgs/lib>; with sources; "${
+    extend
+      (cleanSource ./src)
+      [ ./README.md ]
+  }")'
+)"
+(cd "$dir" && find ..) | sort -f | diff -U10 - <(cat <<EOF
+..
+../README.md
+../src
+../src/bar.c
+EOF
+) || die "sources.extend: can add a file to parent directory"
+
+
+dir="$(
+  nix eval --raw '(with import <nixpkgs/lib>; with sources; "${
+    focusAt ./. (
+      extend
+        (cleanSource ./src)
+        [ ./README.md ]
+    )
+  }")'
+)"
+(cd "$dir" && find .) | sort -f | diff -U10 - <(cat <<EOF
+.
+./README.md
+./src
+./src/bar.c
+EOF
+) || die "sources.focusAt can change to parent directory"
+
+bool="$(
+  nix eval --raw '(with import <nixpkgs/lib>; with sources; "${boolToString (
+     getOriginalFocusPath (
+      focusAt ./. (
+        extend
+          (cleanSource ./src)
+          [ ./README.md ]
+      )
+    ) == ./.
+  )}")'
+)"
+[[ true == "$bool" ]] || die "getOriginalFocusPath example 1"
+
+bool="$(
+  nix eval --raw '(with import <nixpkgs/lib>; with sources; "${boolToString (
+    getOriginalFocusPath (
+      (extend
+        (cleanSource ./src)
+        [ ./README.md ]
+      )
+    ) == ./src
+  )}")'
+)"
+[[ true == "$bool" ]] || die "getOriginalFocusPath example 2"
+
+string="$(
+  nix eval --raw '(with import <nixpkgs/lib>; with sources; "${boolToString (
+    getSubpath (
+      (extend
+        (cleanSource ./src)
+        [ ./README.md ]
+      )
+    ) == "src"
+  )}")'
+)"
+[[ true == "$bool" ]] || die "getSubpath example 1"
```

Comment:
```suggestion
[[ true == "$string" ]] || die "getSubpath example 1"
```
Seems like a typo here?

Reply:
Fixed. Thanks!